### ISSUE TITLE
Support for Beckhoff EL6021

### ISSUE THREAD

**Issue by GuiHue** [linuxcnc-ethercat/linuxcnc-ethercat#294]
Support [Beckhoff EL6021 ](https://www.beckhoff.com/en-us/products/i-o/ethercat-terminals/el6xxx-communication/el6021.html) communication module.

It's a 1-channel communication interface for serial using RS422 and RS485.

Would be a fantastic way to get many VFDs into EtherCAT and not having one of those nasty USB to serial things. ATC router spindles can be a bit tricky since they often have a 4 or more poles which requires VFD capable of driving >400Hz. This limits the choice in the market substantially and often collides with export limitations which mean that non corporate legal entities cannot buy them new. Many established VFDs are not available with EtherCAT (or are stupid expensive).

I have one sitting here.

**Comment by scottlaird** [comment]
I have an EL6001, which I *think* is the same basic hardware, just single-ended instead of differential.  Not that useful for driving VFDs, but easier for testing and probably register-compatible.  IIRC our EtherCAT library has code for EtherCAT-attached serial devices, so it's probably not too hard to code up.

The problem is that I have no idea what they should look like in LinuxCNC.  I don't think that writing a "VFD-brand-X-over-Beckhoff-serial-model-Y-over-EtherCAT" driver is the right way to go :-).  I don't see a "generic serial" or "generic RS485" layer that we could attach into.  The best modbus VFD example that I can see just uses `/dev/ttyXXX` directly.  I guess we *could* present a fake serial port to Linux, and that'd be semi-useful outside of LinuxCNC.  The problem is that I don't *think* there's an actual way to create a serial device from user space.  We could use a PTY, which is close, but you couldn't change baud rate, etc, via ioctls.

Alternately, it looks like there's a `mb2hal` module that can currently talk to both serial and Modbus TCP devices.  I can't remember if Modbus TCP is literally just MB over TCP, or if it's more complicated.  I've only used it once, and it was a few years ago.  We could maybe just listen on a TCP port and feed that through to the EtherCAT serial port?  It'd be kind of weird, structurally, but I don't see why it wouldn't work, and it'd be somewhat flexible.  Of course, then you have issues around networks and security, but they're manageable, mostly.

Do you want RS-422/485, or do you want Modbus/Canbus/etc?

**Comment by scottlaird** [comment]
There is an EL60xx *kernel* driver in https://gitlab.com/etherlab.org/ethercat/-/tree/master/examples/tty.  I'm not sure that we could share a master between this and LCEC, but at least it's an example.

**Comment by grandixximo** [comment]
@GuiHue if you're still active (also pinging you on #435), is the EL6021 still on your bench? This one is genuinely hard because it needs to expose serial to userspace as a tty so existing modbus/VFD components can talk to it. Etherlab has a tty kernel example but it's unclear how to share the master. Lower priority unless you have appetite for a deeper dive.

**Comment by evanwyatt** [comment]
@grandixximo I currently have a EL6021 setup that I'd like to try to connect to a HY VFD. I'm familiar with writing HAL components and keen to give this a try.

I'm torn between just creating a specific EL6021-HY driver just to get it working or trying to implement a generic EL6021 driver that can expose a tty for other components to talk to.

**Comment by grandixximo** [comment]
I think the generic one would by a very good addition, the specific one should run on top of the generic one, this would be useful to anyone in the future, not just for HY VFD

**Comment by evanwyatt** [comment]
It's not possible to share the master, and I'm not sure it's possible to create the tty for user space. 

Current thinking is to expose it over pins to a HAL component and hope the latency isn't too awful. 

I'll ponder more when I've had more coffee.

**Comment by grandixximo** [comment]
IMO the strongest path here is to steal LinuxCNC's own realtime modbus engine rather than invent a tty.

Master sharing is off the table: LCEC owns the EL6021's PDOs via the userspace `ecrt` API, so the etherlab kernel tty example can't attach to the same slave and the serial logic has to live inside an LCEC driver.

The precedent I'd point at is Andy Pugh's `mesa_modbus` (in linuxcnc master, `src/hal/drivers/mesa-hostmot2/modbus`). It's a full realtime modbus RTU master in a HAL comp, no `/dev/tty`, no libmodbus. It builds and parses frames (`build_data_frame`/`parse_data_frame`/`RTU_CRC`) and runs the whole transaction as a cycle-stepped state machine, one step per RT invocation. All of that is transport-agnostic. The only Mesa-specific parts are ~3 functions that shuttle bytes through the PktUART registers.

EL6021 is the same shape: shuttle up to 22 bytes through the PDOs every 2nd cycle, gated by the control/status handshake word. So the suggestion would be a generic `EL6021` LCEC driver that maps the data PDOs + handshake, then swaps those ~3 transport functions to reuse mesa_modbus's framing, CRC, state machine and its `{TYPE, FUNC, ADDR, COUNT, pin_name}` descriptor/pin generation. GPL2 both sides, so lifting the code is clean. This might be the nicest fit: modbus stays fully in the RT cyclic loop, no separate process, and the cycle-stepped state machine maps naturally onto the every-2nd-cycle handshake. It's basically your "expose over pins to a HAL component" idea, but with a proven engine behind it.

If we ever want arbitrary serial (not just modbus), the other option is exposing a real `/dev/ttyECATx` from userspace via CUSE (character-device-in-userspace, libfuse), which can service the termios ioctls itself so existing `/dev/tty` VFD drivers work unmodified. More moving parts (a non-RT bridge), so I'd treat it as the general-purpose escape hatch rather than the first cut.

For #294's actual goal (a modbus VFD), I'd suggest the mesa_modbus port. @evanwyatt what are your thoughts, and would you have appetite to prototype the transport swap against your EL6021 + HY setup?

**Comment by evanwyatt** [comment]
@grandixximo I finally got around to setting up my test bench for this. I have it hooked up to a USB RS485 adapter with a 'fake' HY VFD program listening. So from this point I'm able to play around properly. 

First step is to setup a generic driver `lcec_el6021.c` and map the PDOs, etc.

**Comment by grandixximo** [comment]
Sounds good, if you want to chat join the linuxcnc discord https://discord.gg/s8p9ce8SY
@ me there if you need anything from my end. Or leave a message here, both work, discord is more chat style if you are into that...

**Comment by rodw-au** [comment]
Unrelated but  it may serve as a guide, I just shared a MODBUS RTU configuration for the ethercat catio-a IO module thats available on aliexpress. I have this controlling my VFD now.
Forum: https://forum.linuxcnc.org/ethercat/59180-catio-a-spindle-control-with-modbus-over-ethercat
This device supports MODBUS RTU and custom serial comms. The docs for it are on the forum too but I can share them here if you like. The Google translation from Chinese is not easy to read.

The approach it takes is it has a set of PDOS that data is written to and sent from. Each of these PDO's has a set of SDOS that define the MODBUS register to read from or write to. These SDOs are a bit funny because once they are written to they are read only until a power cycle so I used a bash script to set the environment up before running LinuxCNC.

Once configured, the hal side (which I also shared) it is basically the same as mb2hal (except it uses ints and mb2hal uses floats) so I would recommend you get your VFD going with mb2hal before starting this. I found my VFD could not run higher than 38400 baud so check this with mb2hal to avoid wasting time.

The two parameters that are critical is the MODBUS timeout and the MODBUS cycle time (time between updates). I have not explored in detail but you need to use the fastest baud rate you can given the disparity between the servo thread and MODBUS timings. I did get timeut errors reported but its working perfectly for me now even with these errors. I wondered if the lcec.write_all command in the hal file loop was causing the serial port to flood but have no way to checking this. The device did not work at 9600 baud (well it did but you could not change the frequency ever without really long cycle times that were unuseable).... If the driver could have a way of configuring each register when loaded, (even with a C header file to start with)  your idea might work quite well.
